File: leetcode/tree/2792_Count_Nodes_That_Are_Great_Enough/codes/dfs.cpp
#include <bits/stdc++.h>
using namespace std;

/**
Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    pair<priority_queue<int>, int> dfs(TreeNode* root, int k) {
        if (!root) return {{}, 0};

        auto left = dfs(root->left, k);
        auto right = dfs(root->right, k);

        while (!right.first.empty()) {
            left.first.push(right.first.top());
            right.first.pop();
            if ((int)left.first.size() > k) left.first.pop();
        }

        int count = left.second + right.second;
        if ((int)left.first.size() == k && left.first.top() < root->val) count++;

        left.first.push(root->val);
        if ((int)left.first.size() > k) left.first.pop();

        return {left.first, count};
    }
public:
    int countGreatEnoughNodes(TreeNode* root, int k) {
        return dfs(root, k).second;
    }
};

File: leetcode/tree/2792_Count_Nodes_That_Are_Great_Enough/codes/dfs.py
import heapq
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def countGreatEnoughNodes(self, root: Optional['TreeNode'], k: int) -> int:
        def dfs(node):
            if not node:
                return [], 0
            left_heap, left_count = dfs(node.left)
            right_heap, right_count = dfs(node.right)

            heap = left_heap + right_heap
            heapq.heapify(heap)
            smallest_k = heapq.nsmallest(k, heap)

            count = left_count + right_count
            if len(smallest_k) == k and node.val > smallest_k[-1]:
                count += 1

            heapq.heappush(heap, node.val)
            heap = heapq.nsmallest(k, heap)
            return heap, count

        return dfs(root)[1]
